membership-attribute-service/app/models/Attributes.scala

```diff
@@ -12,60 +10,96 @@ import play.api.mvc.Results.Ok
 }
 
-case class Attributes(
-  UserId: String,
-  Tier: Option[String] = None,
-  MembershipNumber: Option[String] = None,
-  AdFree: Option[Boolean] = None,
-  RecurringContributionPaymentPlan: Option[String] = None,
-  MembershipJoinDate: Option[LocalDate] = None,
-  DigitalSubscriptionExpiryDate: Option[LocalDate] = None,
-  TTLTimestamp: Option[Long] = None
-) {
-
-  require(UserId.nonEmpty)
-
+case class Attributes(UserId: String,
+                      Tier: Option[String] = None,
+                      RecurringContributionPaymentPlan: Option[String] = None,
+                      MembershipJoinDate: Option[LocalDate] = None,
+                      DigitalSubscriptionExpiryDate: Option[LocalDate] = None,
+                      MembershipNumber: Option[String] = None,
+                      AdFree: Option[Boolean] = None) {
   lazy val isFriendTier = Tier.exists(_.equalsIgnoreCase("friend"))
   lazy val isSupporterTier = Tier.exists(_.equalsIgnoreCase("supporter"))
   lazy val isPartnerTier = Tier.exists(_.equalsIgnoreCase("partner"))
   lazy val isPatronTier = Tier.exists(_.equalsIgnoreCase("patron"))
   lazy val isStaffTier = Tier.exists(_.equalsIgnoreCase("staff"))
   lazy val isPaidTier = isSupporterTier || isPartnerTier || isPatronTier || isStaffTier
-  lazy val isAdFree = AdFree.exists(identity)
   lazy val isContributor = RecurringContributionPaymentPlan.isDefined
   lazy val staffDigitalSubscriptionExpiryDate: Option[LocalDate] = Tier.exists(_.equalsIgnoreCase("staff")).option(now.plusDays(1))
   lazy val latestDigitalSubscriptionExpiryDate =  Some(Set(staffDigitalSubscriptionExpiryDate, DigitalSubscriptionExpiryDate).flatten).filter(_.nonEmpty).map(_.max)
   lazy val digitalSubscriberHasActivePlan = latestDigitalSubscriptionExpiryDate.exists(_.isAfter(now))
 
+  lazy val isAdFree = AdFree.exists(identity)
+
   lazy val contentAccess = ContentAccess(member = isPaidTier || isFriendTier, paidMember = isPaidTier, recurringContributor = isContributor, digitalPack = digitalSubscriberHasActivePlan)
 }
 
+case class ZuoraAttributes(UserId: String,
+                           Tier: Option[String] = None,
+                           RecurringContributionPaymentPlan: Option[String] = None,
+                           MembershipJoinDate: Option[LocalDate] = None,
+                           DigitalSubscriptionExpiryDate: Option[LocalDate] = None)
+
+case class DynamoAttributes(UserId: String,
+                            Tier: Option[String] = None,
+                            RecurringContributionPaymentPlan: Option[String] = None,
+                            MembershipJoinDate: Option[LocalDate] = None,
+                            DigitalSubscriptionExpiryDate: Option[LocalDate] = None,
+                            MembershipNumber: Option[String],
+                            AdFree: Option[Boolean],
+                            TTLTimestamp: Long) {
+  lazy val isFriendTier = Tier.exists(_.equalsIgnoreCase("friend"))
```

Comment:
All of this code is duplicated, so we could pull it out of the case classes and into a separate trait or object?

Reply:
Yes, I think there is definitely a better way to model this. Do you have some time to work with me on this point? I thought about it earlier but had some trouble with the Attributes writer.

Reply:
Yes sure - we could always remove this duplication in a separate PR though - it shouldn't block this one as it's not really a GDPR requirement!

Reply:
how about make it a `tier: MaybeTier` with subclasses NoTier,  Friend, Partner, Staff etc and sort out which it is in the readers.  Then the readers can also sort out the ignore caseing so we don't need all this IgnoreCase logic any more.